routing/payment_lifecycle_test.go

```diff
@@ -491,3 +662,943 @@ func TestDecideNextStep(t *testing.T) {
+	m.clock.On("Now").Return(time.Now())
+
+	// We now register attempt and return no error.
+	m.control.On("RegisterAttempt",
+		p.identifier, mock.Anything,
+	).Return(nil).Once()
+
+	// 1.6. mock `sendPaymentAttempt` to succeed, which brings us into the
+	// next iteration of the lifecycle.
+	m.payer.On("SendHTLC",
+		mock.Anything, attemptID, mock.Anything,
+	).Return(nil).Once()
+
+	// We now enter the second iteration of the lifecycle loop.
+	//
+	// 2.1. calls `FetchPayment` and return the payment.
+	m.control.On("FetchPayment", p.identifier).Return(m.payment, nil).Once()
+
+	// 2.2. calls `GetState` and return the state.
+	m.payment.On("GetState").Return(ps).Run(func(args mock.Arguments) {
+		ps.RemainingAmt = 0
+	}).Once()
+
+	// 2.3. decideNextStep now returns stepExit and exits the loop.
+	m.payment.On("AllowMoreAttempts").Return(false, nil).Once().
+		On("NeedWaitAttempts").Return(false, nil).Once()
+
+	// We should perform an optional deletion over failed attempts.
+	m.control.On("DeleteFailedAttempts", p.identifier).Return(nil).Once()
+
+	// Finally, mock the `TerminalInfo` to return the settled attempt.
+	// Create a SettleAttempt.
+	testPreimage := lntypes.Preimage{1, 2, 3}
+	settledAttempt := makeSettledAttempt(t, int(paymentAmt), testPreimage)
+	m.payment.On("TerminalInfo").Return(settledAttempt, nil).Once()
+
+	// Send the payment and assert the preimage is matched.
+	sendPaymentAndAssertSucceeded(t, p, testPreimage)
+
+	// Expected collectResultAsync to called.
+	require.Equal(t, 1, m.collectResultsCount)
+}
+
+// TestResumePaymentRetry checks a payment flow with a failure on the first
+// attempt, but that succeeds on the second attempt.
+//
+// NOTE: No parallel test because it overwrites global variables.
+//
+//nolint:paralleltest
+func TestResumePaymentRetry(t *testing.T) {
+	// Create a test paymentLifecycle with the initial two calls mocked.
+	p, m := setupTestPaymentLifecycle(t)
+
+	// Create a dummy route that will be returned by `RequestRoute`.
+	paymentAmt := lnwire.MilliSatoshi(10000)
+	rt := createDummyRoute(t, paymentAmt)
+
+	// We now enter the payment lifecycle loop.
+	//
+	// 1.1. calls `FetchPayment` and return the payment.
+	m.control.On("FetchPayment", p.identifier).Return(m.payment, nil).Once()
+
+	// 1.2. calls `GetState` and return the state.
+	ps := &channeldb.MPPaymentState{
+		RemainingAmt: paymentAmt,
+	}
+	m.payment.On("GetState").Return(ps).Once()
+
+	// NOTE: GetStatus is only used to populate the logs which is
+	// not critical so we loosen the checks on how many times it's
+	// been called.
+	m.payment.On("GetStatus").Return(channeldb.StatusInFlight)
+
+	// 1.3. decideNextStep now returns stepProceed.
+	m.payment.On("AllowMoreAttempts").Return(true, nil).Once()
+
+	// 1.4. mock requestRoute to return an route.
+	m.paySession.On("RequestRoute",
+		ps.RemainingAmt, p.feeLimit, uint32(ps.NumAttemptsInFlight),
+		uint32(p.currentHeight),
+	).Return(rt, nil).Once()
```

Comment:
note that the returned route can also be considered as mocked. And we should not test the returned route because it's the `RequestRoute`'s responsibility to make sure it's correct.

That being said, `rt` is not perfectly mocked here yet as we need to interface `*route.Route` to make it more testable.